### ISSUE TITLE
Adjustable Precision (Microsecond) in BLE Payload Offsets

### ISSUE THREAD

**Issue by blainerothrock** [Flexi-BLE/flexible-swift#2]
We need the ability to record timestamps in nanosecond precision. This can be completed in two steps. The first is to support nanosecond offsets in the BLE payloads (still only recording millisecond precision in the reference time). 

An additional issue will be created to support Micosecond for the reference time. It may not be required for a while.

**Comment by blainerothrock** [comment]
Fixed in 0.3 (4)

**Comment by blainerothrock** [comment]
We are able to add this feature to both the anchor timestamp and data stream offsets but run into a few complexities, which we handle accordingly: 

### Indicating Prevision
* a new field in the JSON specification: `precision`, accepts the values `ms` (millisecond) or `us` (microsecond).

### Database storage
UINT64 numbers cannot store microsecond precision (requires 16 digits for UNIX epoch time), therefore date objects cannot handle this, both Swift and SQLite. For convenience, we still want to store date objects in the database, so to fix this, when precision is set to microseconds, a new column is added to the database `ts_precision` (integer) which stores the fractional part in seconds. 
* **NOTE**: When plotting at millisecond precision we will need to use indices only to refer to timestamps.
* **NOTE**: This will also support nanosecond precision, although I cannot see how we would use it.

### Update to Reference Time
A UINT32 system uptime on a peripheral microcontroller would not overflow for ~49 days when dealing in milliseconds. With microsecond precision, this is reduced to ~71 minutes. Therefore we need to now handle requesting the reference time to re-sync when the system uptime is about to overflow. We accomplish this by creating a new info characteristic (notify) which flips from 0 to 1 when the peripheral requires a new reference time sync. At this point, the application writes the reference time again.

```mermaid
sequenceDiagram
    participant A as App
    participant P as Peripheral
    A->>+P: BLE Connection Established
    A->>+P: Write Reference Time
    P->>P: Stores System Time
    Note over A,P: Data Collected Over Time ...
    P->>A: Datastream BLE Notify
    P->>P: Detects Overflow in System Time
    P->>A: Notifies (1) on Refernce Time Request
    A->>P: Writes Reference Time
    P->>P: Stores System Time
    Note over A,P: Data Collection Continues as Configured
```
